### PR TITLE
GCP CLI support: `tsh` and `tctl` changes

### LINKED ISSUE #17257: GCP API access
## Description

Teleport currently provides a way to log users in AWS console using federation API as well as access to AWS API via `tsh aws` and `tsh proxy aws` commands:

https://goteleport.com/docs/application-access/guides/aws-console/

Accessing Azure console/API is currently in progress.

We would like to add support for similar functionality for accessing GCP.

## Success criteria

- [ ] Teleport provides `tsh gcp` command users can use to interact with GCP API.
- [ ] Users can start local GCP proxy with `tsh proxy gcp` which GCP API clients can use (e.g. terraform).

# Next steps

Console access:

- [ ] Users can register a GCP dashboard in app access and connect to it with a selected user/role.
  - ~We need to investigate if GCP provides similar kind of functionality as [AWS federation API](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_enable-console-custom-url.html) that would allow Teleport to generate a sign-in link.~
  - As far as we know this isn't possible. We will be able to support this once https://github.com/gravitational/teleport/issues/18043 is delivered.

## Useful references

* Umbrella ticket for AWS/Azure/GCP access: https://github.com/gravitational/teleport/issues/11780
* Azure access ticket (no PRs yet): https://github.com/gravitational/teleport/issues/16401
* AWS console access PR: https://github.com/gravitational/teleport/pull/7590
* AWS CLI access PR: https://github.com/gravitational/teleport/pull/8151

### PR DESCRIPTION
**Pull request by Tener**
`tsh` and `tctl` changes for https://github.com/gravitational/teleport/issues/17257

To be applied on top of https://github.com/gravitational/teleport/pull/19789.

This is the final PR in the chain. Fixes https://github.com/gravitational/teleport/issues/17257.

---

**TODO:** Testing: 
- [x] Google Storage interaction with `gsutil` [tool](https://cloud.google.com/storage/docs/gsutil). PR: https://github.com/gravitational/teleport/pull/20072
- [x] Terraform GCP provider. This works out of the box after configuring Terraform.

---

Setup:
1. Create a service account `teleport-service`. 
2. Assign `Service Account User` role to the `teleport-service` account.
3. Deploy Teleport instance to GCP VM. Select `teleport-service` as a VM account to use. If you cannot select the service account, check that it has appropriate role (see step 2).
4. Create service accounts for users to use, for example:`teleport-user-1` and `teleport-user-2`. Assign desired permissions for these accounts. 
5. Grant access for `teleport-user-1`, `teleport-user-2` etc. to `teleport-service` at least with `Service Account Token Creator` role. This will allow `teleport-service` to issue tokens as `teleport-user-XXX`. More info: https://cloud.google.com/iam/docs/service-accounts#token-creator-role 
6. Configure GCP as a Teleport application. `cloud: GCP` is the critical part of the config.
```yaml
app_service:
  enabled: "yes"
  apps:
  - name: gcpapp
    cloud: GCP
```
7. Ensure the user is allowed to use chosen service accounts. If this is a fresh cluster, the `access` role should contain the snippet below. If not, modify a role the user has access to include a reference to `{{internal.gcp_service_accounts}}`. You can also hardcode the list of identities in the role itself (not recommended).
```yaml
spec:
  allow:
    gcp_service_accounts:
    - '{{internal.gcp_service_accounts}}'
```
10. When using `{{internal.gcp_service_accounts}}`, the list of allowed roles is sourced from user-specific traits. To update an existing user use `tctl users update <username> --set-gcp-service-accounts ID_1,ID_2,ID_3,...`, where `ID_N` is a full name of service account (e.g. `teleport-user-1@my-account-123456.iam.gserviceaccount.com` )
11. Refresh client login with `tsh logout; tsh login --proxy=...`.
12. Log in to `gcpapp` app with `tsh app login gcpapp [--gcp-service-account ACCOUNT]`. The service account is optional if there is just one available. The service account name can be shortened by discarding domain (e.g. `teleport-user-1`) as long as it doesn't make it ambigous.
13. Run commands with `tsh gcloud ...`, e.g. `tsh gcloud compute instances list`. 
14. Alternatively, start a proxy with `tsh proxy gcloud` and follow printed instructions.

Notes:
1. Instead of deploying Teleport to GCP VM, you can download the key for `teleport-service` service account to your local machine and set the `GOOGLE_APPLICATION_CREDENTIALS` env variable to its location. This is handful in development, but not recommended for production as the key file can be stolen. Example: `GOOGLE_APPLICATION_CREDENTIALS=/path/to/teleport-service-my-account-123456-ee0011223344.json teleport start`
2. The setup above assumes Teleport is running as a privileged account `teleport-service` which is inaccessible to Teleport users.

### CODE REVIEW

File: lib/srv/alpnproxy/gcp_middleware.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+func (m *GCPMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
+	auth := req.Header.Get("Authorization")
+	if auth == "" {
+		m.Log.Debugf("No Authorization header present, ignoring request.")
+		return false
+	}
+
+	expectedAuth := "Bearer " + m.Secret
+
+	if auth != expectedAuth {
```
Comment (GavinFrazar) [inline lib/srv/alpnproxy/gcp_middleware.go:54]: This check is in the local proxy, so it probably doesn't *really* matter, but we should do constant time comparison:
```
subtle.ConstantTimeCompare([]byte(auth), []byte(expectedAuth))
```

File: tool/tsh/gcp.go [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+}
+
+// getGCPSecret will return fresh secret to use or read it from environment.
+func getGCPSecret() (string, error) {
+	secret := os.Getenv("TELEPORT_GCP_SECRET")
+	if secret != "" {
+		return secret, nil
+	}
+
+	return utils.CryptoRandomHex(10)
```
Comment (GavinFrazar) [inline tool/tsh/gcp.go:95]: Why 10? Elsewhere we use CryptoRandomHex I see we use at least 16 and up to 32, but idk if it actually matters here. We have `TokenLenBytes` defined in lib/auth/auth.go as 16 as well.

File: tool/tsh/gcp.go [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+		cf:      cf,
+		profile: profile,
+		app:     app,
+		secret:  secret,
+	}, nil
+}
+
+// getGCPSecret will return fresh secret to use or read it from environment.
+func getGCPSecret() (string, error) {
+	secret := os.Getenv("TELEPORT_GCP_SECRET")
```
Comment (GavinFrazar) [inline tool/tsh/gcp.go:90]: I think we should make this a constant in constants.go with other env vars we define

File: tool/tsh/gcp.go [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+func (a *gcpApp) GetEnvVars() (map[string]string, error) {
+	envVars := map[string]string{
+		// TODO: verify secret in calls somehow
+		"CLOUDSDK_AUTH_ACCESS_TOKEN": a.secret,
+
+		// Needed for gcloud CLI to accept our certs.
+		// This isn't portable and applications may have to set different env variables,
+		// add the application cert to system root store (not recommended, ultimate fallback)
+		// or use equivalent of --insecure flag.
+		"REQUESTS_CA_BUNDLE": a.profile.AppLocalCAPath(a.app.Name),
```
Comment (GavinFrazar) [inline tool/tsh/gcp.go:144]: nit: add reference URL link to documentation for these env vars if possible

File: tool/tsh/gcp.go [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+func (s SortedGCPServiceAccounts) Less(i, j int) bool {
+	beforeI, afterI, _ := strings.Cut(s[i], "@")
+	beforeJ, afterJ, _ := strings.Cut(s[j], "@")
+
+	if afterI != afterJ {
+		return afterI < afterJ
+	}
+
+	return beforeI < beforeJ
+}
```
Comment (GavinFrazar) [inline tool/tsh/gcp.go:323]: Could you add docs describing the expected behavior of this sorting? I think it's sorting first by what follows @ and then by what precedes "@" if the suffix is equal?

File: tool/tsh/gcp.go [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+		if app.GCPServiceAccount == "" {
+			return nil, trace.BadParameter(
+				"Selected app %q is not an GCP application", name,
+			)
+		}
+		return newGCPApp(cf, profile, *app)
+	}
+	gcpApps := getGCPApps(profile.Apps)
+	if len(gcpApps) == 0 {
+		return nil, trace.NotFound("Please login to GCP App using 'tsh app login' first")
```
Comment (GavinFrazar) [inline tool/tsh/gcp.go:421]: nit: 
```suggestion
		return nil, trace.NotFound("Please login to a GCP App using 'tsh app login' first")
```

**Review by GavinFrazar — CHANGES_REQUESTED** [review]
Mostly LGTM, but a few things I'd like to address before approving

File: tool/tsh/app.go [hunk 7]
```diff
@@ -200,6 +219,14 @@ Your identity: {{.identity}}
 Example Azure CLI command: tsh az vm list
 `))
 
+// gcpCliTpl is the message that gets printed to a user upon successful login
+// into a GCP application.
+var gcpCliTpl = template.Must(template.New("").Parse(
+	`Logged into GCP app "{{.appName}}".
+Your service account: {{.serviceAccount}}
+Example command: tsh gcloud compute instances list
```
[the remark this answers is not in the document]
Comment (Tener) [reply tool/tsh/app.go:227]: `gcloud` is the nominal name of the CLI binary. Without `tsh` the command `gcloud compute instances list` would work, whereas `gcp compute instances list` would not. I didn't realize it doesn't show up in `tsh -h`. I'll make `gcloud` the main command and `gcp` the alias.

File: tool/tsh/gcp.go [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+	// // backend expects the tokens to be signed with web session private key
+	// ws, err := tc.GetAppSession(a.cf.Context, types.GetAppSessionRequest{SessionID: a.app.SessionID})
+	// if err != nil {
+	// 	return trace.Wrap(err)
+	// }
+
+	// wsPK, err := utils.ParsePrivateKey(ws.GetPriv())
+	// if err != nil {
+	// 	return trace.Wrap(err)
+	// }
```
[the remark this answers is not in the document]
Comment (Tener) [reply tool/tsh/gcp.go:223]: Ah, indeed. Thanks.

File: tool/tsh/gcp.go (hunk 2 above)
Comment (Tener) [reply tool/tsh/gcp.go:95]: `auth.TokenLenBytes` is a good default, I'll use it instead.

File: tool/tsh/gcp.go (hunk 5 above)
Comment (Tener) [reply tool/tsh/gcp.go:323]: Good point, will do.

File: lib/srv/alpnproxy/gcp_middleware.go (hunk 1 above)
Comment (Tener) [reply lib/srv/alpnproxy/gcp_middleware.go:54]: Huh. I wonder how practical such an attack is, but this is good practice indeed.

File: tool/tsh/gcp.go (hunk 3 above)
Comment (Tener) [reply tool/tsh/gcp.go:90]: `tsh` has a bunch of env vars defined for its own use only, it is rightfully in `tsh.go`. I'll add an entry there.

File: tool/tsh/gcp.go (hunk 4 above)
Comment (Tener) [reply tool/tsh/gcp.go:144]: Trying to find the docs for `REQUESTS_CA_BUNDLE` I realized it is non-standard env var and isn't documented anywhere. Instead, there is [`core.custom_ca_certs_file`](https://cloud.google.com/sdk/gcloud/reference/config/set#:~:text=custom_ca_certs_file) property which can be set via env variables too (`CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE`).

File: lib/srv/alpnproxy/gcp_middleware.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	"crypto/subtle"
+	"net/http"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// GCPMiddleware implements a simplified version of MSI server serving auth tokens.
```
Comment (smallinsky) [inline lib/srv/alpnproxy/gcp_middleware.go:25]: I think that **MSI** is Azure specific name and doesn't apply to Google Cloud

File: lib/srv/alpnproxy/gcp_middleware.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+type GCPMiddleware struct {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/alpnproxy/gcp_middleware.go:26]: Yeah, I guess this isn't really GCP specific anymore.

File: lib/srv/alpnproxy/gcp_middleware.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+		m.Log = logrus.WithField(trace.Component, "gcp")
+	}
+
+	if m.Secret == "" {
+		return trace.BadParameter("missing Secret")
+	}
+	return nil
+}
+
+func (m *GCPMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/alpnproxy/gcp_middleware.go:46]: Middleware can do much more than just return an error. For Azure this is where MSI is implemented: https://github.com/gravitational/teleport/blob/3ebbfdd222eb5111aebe17d079d694d6c08776ec/lib/srv/alpnproxy/azure_msi_middleware.go#L81-L133

`HandleRequest` can be seen as an event handler. Returning a `bool` from event handler to indicate if the event was handled or not is a pretty standard convention.

[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/alpnproxy/gcp_middleware.go]: Docs added.

File: lib/srv/alpnproxy/gcp_middleware.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/alpnproxy/gcp_middleware.go]: Renamed, added comments.

File: lib/srv/alpnproxy/gcp_middleware.go (hunk 9 above)
Comment (Tener) [reply lib/srv/alpnproxy/gcp_middleware.go]: Fixed docs here and elsewhere.

File: lib/srv/alpnproxy/gcp_middleware.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+type GCPMiddleware struct {
+	// Log is the Logger.
+	Log logrus.FieldLogger
+	// Secret to be provided by the client.
+	Secret string
+}
+
+var _ LocalProxyHTTPMiddleware = &GCPMiddleware{}
+
+func (m *GCPMiddleware) CheckAndSetDefaults() error {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/alpnproxy/gcp_middleware.go]: Added.

**Review by GavinFrazar — APPROVED** [review]

File: tool/tsh/gcp.go [hunk 13]
```diff
@@ -0,0 +1,439 @@
...
+// StartLocalProxies sets up local proxies for serving GCP clients.
+//
+// At minimum clients should work with these variables set:
+// - HTTPS_PROXY, for routing the traffic through the proxy
+// - MSI_ENDPOINT, for informing the client about credential provider endpoint
+//
+// The request flow to remote server (i.e. GCP APIs) looks like this:
+// clients -> local forward proxy -> local ALPN proxy -> remote server
+//
+// However, with MSI_ENDPOINT variable set, clients will reach out to this address for tokens.
```
Comment (greedy52) [inline tool/tsh/gcp.go:108]: nit: update doc

**Review by greedy52 — APPROVED** [review]
LGTM. And great setup guide 👍

**Review by smallinsky — APPROVED** [review]

**Comment by Tener** [comment]
@smallinsky 

Regarding point:
> - [x] Terraform GCP provider.

This works out of the box, but Terraform needs to be configured correctly.

Assuming we are running `tsh proxy gcp` like so:

```
TELEPORT_GCLOUD_SECRET=my_gcloud_secret tsh proxy gcp --port 3500
Started GCP proxy on http://127.0.0.1:3500.

Use the following credentials and HTTPS proxy setting to connect to the proxy:

  export BOTO_CONFIG=/Users/myuser/.tsh/gcp/teleport.example.com/gcp/bd460f17_boto.cfg
  export CLOUDSDK_AUTH_ACCESS_TOKEN=my_gcloud_secret
  export CLOUDSDK_CONFIG=/Users/myuser/.tsh/gcp/teleport.example.com/gcp/gcloud
  export CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE=/Users/myuser/.tsh/keys/teleport.example.com/myuser-app/teleport.example.com/gcp-localca.pem
  export CLOUDSDK_CORE_PROJECT=myproject-123456
  export HTTPS_PROXY=http://127.0.0.1:3500
```

Then we must run Terraform with the following env variables exported:

```
  export HTTPS_PROXY=http://127.0.0.1:3500
  export GOOGLE_OAUTH_ACCESS_TOKEN=my_gcloud_secret
```

Instead of using `GOOGLE_OAUTH_ACCESS_TOKEN` the token can be saved in a file as per https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#access_token. 

Unlike some other apps, Terraform does not provide explicit support for custom CA certs: https://github.com/hashicorp/terraform/issues/28551. 

To make it trust `/Users/myuser/.tsh/keys/teleport.example.com/myuser-app/teleport.example.com/gcp-localca.pem` we must use platform-specific mechanisms available to Go-based programs. On Linux that can be `SSL_CERT_FILE` env var, on MacOS `security add-trusted-cert`. I'm not sure what the preferred mechanism is on Windows.

**Pull request merged**